### PR TITLE
[EC-302] Adding maximum gasLimit validation

### PR DESCRIPTION
**Pull request by KonradStaniec**
Validating that block header gas limit is lower than 2**63-1

### CODE REVIEW

**Comment by rtkaczyk** [comment]
Could we add a unit test for this?

File: src/main/scala/io/iohk/ethereum/validators/BlockHeaderValidator.scala [hunk 1]
```diff
@@ -101,15 +102,21 @@ class BlockHeaderValidatorImpl(blockchainConfig: BlockchainConfig) extends Block
...
     * @param blockHeader BlockHeader to validate.
     * @param parentHeader BlockHeader of the parent of the block to validate.
     * @return BlockHeader if valid, an [[HeaderGasLimitError]] otherwise
     */
   private def validateGasLimit(blockHeader: BlockHeader, parentHeader: BlockHeader): Either[BlockHeaderError, BlockHeader] = {
-    val gasLimitDiff = (blockHeader.gasLimit - parentHeader.gasLimit).abs
-    val gasLimitDiffLimit = parentHeader.gasLimit / GasLimitBoundDivisor
-    if(gasLimitDiff < gasLimitDiffLimit && blockHeader.gasLimit >= MinGasLimit) Right(blockHeader)
-    else Left(HeaderGasLimitError)
+    if(blockHeader.gasLimit > MaxGasLimit)
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/validators/BlockHeaderValidator.scala]: style: space after `if`

File: src/main/scala/io/iohk/ethereum/validators/BlockHeaderValidator.scala [hunk 2]
```diff
@@ -101,15 +102,21 @@ class BlockHeaderValidatorImpl(blockchainConfig: BlockchainConfig) extends Block
...
+      Left(HeaderGasLimitError)
+    else {
+      val gasLimitDiff = (blockHeader.gasLimit - parentHeader.gasLimit).abs
+      val gasLimitDiffLimit = parentHeader.gasLimit / GasLimitBoundDivisor
+      if(gasLimitDiff < gasLimitDiffLimit && blockHeader.gasLimit >= MinGasLimit) Right(blockHeader)
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/validators/BlockHeaderValidator.scala]: style: space after `if`

Comment (rtkaczyk) [reply src/main/scala/io/iohk/ethereum/validators/BlockHeaderValidator.scala]: It would be more concise to just extend this conditional with `MaxGasLimit`

Comment (KonradStaniec) [reply src/main/scala/io/iohk/ethereum/validators/BlockHeaderValidator.scala]: I did not want to calculate unnecessary `gasLimitDiff` and `gasLimitDiffLimit`, but this is probably such small thing that for me both versions are ok

**Comment by ntallar** [comment]
Code LGTM!
Do you know if Geth/Parity implement this EIP or if it is implemented in Ethereum Classic? I know it's needed for passing the tests but I wasn't able to find any information on when it was accepted and included into ETH/ETC.

**Comment by KonradStaniec** [comment]
It is implemented definitely in geth and pytheruem.
https://github.com/ethereum/pyethereum/blob/8d69d808096fb599b8b7b6716c1b3e8df6d4a307/ethereum/blocks.py#L449
https://github.com/ethereum/go-ethereum/blob/27a5622e995d762683dd1a79423d83fcf3e62ccf/consensus/ethash/consensus.go#L247

Also under ticket ( https://github.com/ethereum/EIPs/issues/106 ) it seems that java and c++ client also have it.

**Comment by KonradStaniec** [comment]
After research added a possibility to configure eip106 block number similar way to parity. If the `eip106-block-number` is not defined, it will have default value `Long.maxValue` so basically it will be unused.

File: src/main/scala/io/iohk/ethereum/utils/Config.scala [hunk 3]
```diff
@@ -280,6 +281,8 @@ object BlockchainConfig {
       }
 
       override val monetaryPolicyConfig = MonetaryPolicyConfig(blockchainConfig.getConfig("monetary-policy"))
+
+      override val eip106BlockNumber: BigInt = Try(blockchainConfig.getString("eip106-block-number")).map(BigInt(_)).getOrElse(Long.MaxValue)
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/utils/Config.scala]: I'd rather avoid optional settings in the config. Can we make this required and use high values for disabling? Just like other clients do.

Comment (KonradStaniec) [reply src/main/scala/io/iohk/ethereum/utils/Config.scala]: Actually parity uses optional values https://github.com/paritytech/parity/blob/246b5282e546333c086b2e725de024fdfab0b8d9/json/src/spec/ethash.rs#L101, but ok that way we will be consistent with our other configuration values.

**Comment by rtkaczyk** [comment]
Please copy the new setting (commented out) with doc to `src/universal/conf/blockchain.conf`. Those files serve as config documentation for the users.

**Pull request merged**